c++/src/kj/common.h

```diff
@@ -1377,6 +1377,25 @@ class Maybe {
     }
   }
 
+  template <typename F, typename RetOk = EnableIf<canConvert<T, decltype(instance<F>()())>()>>
+  T orDefault(F&& lazyDefaultValue) const & {
```

Comment:
Hmm, looking closer at this, I guess it's sort of awkward that the `const &` version of this returns a `T` and not `const T&` (which is what the non-lazy version returns). This means this method will only work when `T` is a type that has a const copy constructor.

But if you change this to return `const T&`, then that requires that the callback function returns exactly `const T&`, which may be inconvenient in cases where `T` is in fact const-copyable (e.g. `Maybe<int>`).

I wonder if what we really need here is to let the compiler deduce an appropriate type, the same way it would for a ternary operator. That is:

```
template <typename F, typename Result = decltype(instance<bool>() ? instance<const T&>() : instance<F>()())>
Result orDefault(F&& lazyDefaultValue) const & {
```

I think in this case, if the lazy callback does return a `const T&`, then the compiler will deduce the result as `const T&`, but otherwise it'll figure out an appropriate middleground that both expressions can be cast to?

On another note, you are missing an overload for non-const &, which is important to cover due to the enforcement of transitive constness in KJ. You should probably also cover `const &&` for completeness, though admittedly that case rarely comes up in practice.

Reply:
The reason I omitted the `T&` was precisely for the reason of returning `T&`/`const T&`. I'll play around with this & see if what your recommending works out without any dangerous edge cases.

Reply:
I did the change & it seems to work. I'm a little paranoid that I may have missed a test case of some kind but I can't think what it might be. I spot checked some usages that should be illegal & they resulted in compilation errors but hard to know if I was thorough & those are always annoying to write unit tests for.